### PR TITLE
Set compiler flag 'unwind_info' to false if specified in config

### PR DESCRIPTION
**Pull request by TimonPost**
Regarding issue #6541, it was reported that unloading a large module on macOS can result in significantly long loading times. One suggestion was to disable unwind frames, but this didn't seem to have any effect. Upon further investigation, it appears that this issue may be a regression introduced in a commit made last year in August. The specific commit can be found here: [link to commit](https://github.com/bytecodealliance/wasmtime/pull/4643/files#diff-908a422cef87524478865a23918af79040c970e02a3cc6862f6eb508710606f9L1436).

Currently, the config flag is enabled by default, and there is no option to turn it off via the `native_unwind_info` function. This pull request addresses the issue.

I would like to have it confirmed if this means that all crane lift users have been forced to compile with unwind frames, always, or perhaps I am missing something.

Also, second question, disabling this flag, it would only affect GC tracing and debugging right? So should be fine to expose this to users if they use neither of those?  

As temporary workaround i can do:

```
unsafe {
    config.native_unwind_info(false);
    config.cranelift_flag_set("unwind_info", "false");
}
```

### CODE REVIEW

**Comment by jameysharp** [comment]
I think it might be important to make the `native_unwind_info` field be an `Option<bool>` or similar, so that if nobody calls `Config::native_unwind_info` then we know we can set it whichever way makes sense.

Also, I think the existing error messages, about unwind info being necessary for profiling, might be misleading. On Linux, `perf` defaults to frame-pointer unwinding, which I believe is sufficient today on x86 since we always produce frame pointers. We also have a Cranelift setting to force frame pointers, which we could use when profiling instead of registering unwind info. I think Nick said we'll need more sophisticated unwinding once we have exception-handling support though.

In short, I think we need to sort out exactly when native unwind info is actually necessary before reviewing this PR in detail, because I think the overall logic may not be right.

**Comment by jameysharp** [comment]
I think it makes sense to discuss when native unwind info is necessary separately from #6541, which is about native unwind info being slow. But I don't think we need to open a new issue for it; this PR is a fine place to sort out the details, in my opinion.

I'm hoping both @bjorn3 and @fitzgen can give more context on what the native unwind info is actually used for.

**Comment by fitzgen** [comment]
I tested locally, and we do indeed produce identical binaries, both with `.eh_frame` (I'm on Linux) regardless whether this option is set to `true` or `false`. Whoops! We should definitely fix this!

--------------------

> Also, I think the existing error messages, about unwind info being necessary for profiling, might be misleading. On Linux, `perf` defaults to frame-pointer unwinding, which I believe is sufficient today on x86 since we always produce frame pointers. We also have a Cranelift setting to force frame pointers, which we could use when profiling instead of registering unwind info.

A couple small corrections / refinements:

1. `perf` defaults to frame pointers but doesn't always use them. In fact, there are `libc` asm functions that often end up in Wasmtime profiles that don't preserve frame pointers IIRC but do have CFI pragmas, and `perf` will only capture these correctly when you're using DWARF stack walking. (I may have this inverted, though). In any case, assuming that `perf` is always in frame pointers mode is a no go.

2. Wasmtime always forces the compiler to emit frame pointers (to enable fast+simple stack walking) regardless of platform. (The Cranelift x86 backend will always emit frame pointers regardless of the Cranelift setting to preserve them or not.)

So with (1), the message saying native unwind info is *required* for profiling is probably a bit strong, but it also isn't *not* required. It is subtle, and emitting the unwind info in that case, even if it doesn't end up used by the profiler's exact configuration, doesn't hurt so I'd like to continue to emit native unwind info whenever profiling support is configured.

------------

Backing up a bit: the reason we enable native unwind info by default is to be good citizens of the platform and do our best to interoperate with external tooling (such as profilers or debuggers or whatever) that are usually expecting to see the platform's unwind info. We don't rely on it for correctness, which is why we can (modulo bugs) turn it off. I think the good-citizen angle is still well-motivated, so I don't want to change defaults here.

**Comment by fitzgen** [comment]
`Option<bool>` and then reading it something like `self.native_unwind_info.unwrap_or(DEFAULT_NATIVE_UNWIND_INFO)` makes sense to me.

File: crates/wasmtime/src/config.rs [hunk 1]
```diff
@@ -1600,7 +1605,23 @@ impl Config {
...
+            if self.profiling_strategy == ProfilingStrategy::None {
+                if let Strategy::Cranelift = self.compiler_config.strategy {
+                    unsafe {
+                        self.cranelift_flag_set("unwind_info", "false");
+                    }
+                }
+            } else {
+                bail!(
+                    "compiler option 'unwind_info' can not be disabled when profiling is enabled"
+                );
```
Comment (fitzgen) [inline crates/wasmtime/src/config.rs:1624]: Yeah, I think it would be slightly nicer to do something like

```rust
if self.profiling_strategy != ProfilingStrategy::None ||
    target == target_lexicon::OperatingSystem::Windows
{
    self.cranelift_flag_set("unwind_info", "true");
}

if !self.compiler_config.ensure_setting_unset_or_given(
    "unwind_info", 
    self.native_unwind_info.unwrap_or(DEFAULT_UNWIND_INFO),
) {
    bail!("`native_unwind_info` cannot be disabled when profiling or on Windows");
}
```

Comment (TimonPost) [reply crates/wasmtime/src/config.rs:1624]: - If profiling enabled, always include unwind info, and ignore `self.native_unwind_info`, even if set by user. We likely want to add this to the functoin comments?
- I assume `DEFAULT_UNWIND_INFO` to true?

**Comment by alexcrichton** [comment]
I might have missed some context reading over this, but how come the profiling strategy is consulted when determining whether to enable this? I know that jitdump/perfmap/guest don't use native unwinding information (`perf` can't find it even if we emit it), although I don't know about vtune. Given that I don't think we should force enable the unwinding information if profiling is enabled, instead sticking to whatever the default was.

**Comment by fitzgen** [comment]
> I might have missed some context reading over this, but how come the profiling strategy is consulted when determining whether to enable this? I know that jitdump/perfmap/guest don't use native unwinding information (`perf` can't find it even if we emit it), although I don't know about vtune. Given that I don't think we should force enable the unwinding information if profiling is enabled, instead sticking to whatever the default was.

Ah, I must be operating on outdated information (just like the configuration code itself must have been) because I was under the impression that `perf` used this information when configured to consume DWARF.

If not, then yeah, we can probably decouple profiling and native unwind info.

**Review by fitzgen — COMMENTED** [review]
LGTM, modulo the new revelations around profiling. Thanks!

File: crates/wasmtime/src/config.rs [hunk 2]
```diff
@@ -1579,7 +1579,12 @@ impl Config {
             .insert("probestack_strategy".into(), "inline".into());
 
         let host = target_lexicon::Triple::host();
-        let target = self.compiler_config.target.as_ref().unwrap_or(&host);
+        let target = self
+            .compiler_config
+            .target
+            .as_ref()
+            .unwrap_or(&host)
+            .clone();
```
Comment (jameysharp) [inline crates/wasmtime/src/config.rs:1587]: I don't think this `.clone()` is necessary after the various changes this PR has gone through, though I could be wrong.

**Review by jameysharp — COMMENTED** [review]
Okay, I think we have consensus on the conditions where this setting is needed, and now just need to agree on implementation.

**Comment by jameysharp** [comment]
Hey @fitzgen, the description of Cranelift's `unwind_info` setting says:

>This increases metadata size and compile time, but allows for the debugger to trace frames, is needed for GC tracing that relies on libunwind (such as in Wasmtime), and is unconditionally needed on certain platforms (such as Windows) that must always be able to unwind.

Does Wasmtime GC still require libunwind? I assumed the fast stack walking is used there now.

I'm wondering if we should make this Cranelift setting default to "false" at some point.

File: crates/wasmtime/src/config.rs [hunk 3]
```diff
@@ -1592,18 +1597,25 @@ impl Config {
...
+        const DEFAULT_UNWIND_INFO: bool = true;
+        if !self.compiler_config.ensure_setting_unset_or_given(
+            "unwind_info",
+            &self
+                .native_unwind_info
+                .unwrap_or(DEFAULT_UNWIND_INFO)
+                .to_string(),
+        ) {
+            bail!("`native_unwind_info` cannot be disabled when profiling or on Windows");
+        }
```
Comment (jameysharp) [inline crates/wasmtime/src/config.rs:1617]: Given the new information that profiling doesn't affect this setting at all, I think the structure should be more like:

```rust
if let Some(unwind_requested) = self.native_unwind_info {
    if !self.compiler_config.ensure_setting_unset_or_given("unwind_info", unwind_requested.to_string()) {
        bail!("incompatible settings requested for Cranelift and Wasmtime unwind-info settings");
    }
}

if target.operating_system == target_lexicon::OperatingSystem::Windows {
    if !self.compiler_config.ensure_setting_unset_or_given("unwind_info", "true") {
        bail!("`native_unwind_info` cannot be disabled on Windows");
    }
}
```

**Review by jameysharp — COMMENTED** [review]
I failed to submit one of my review comments, whoops?

**Comment by bjorn3** [comment]
> I'm wondering if we should make this Cranelift setting default to "false" at some point.

That would break panic backtraces. The only way wasmtime can influence panic backtrace printing is by registering unwindinfo. It can't hook unwinding.

**Comment by bjorn3** [comment]
I mean in wasmtime. And there not the unwinding, but the unwinding to get a backtrace as printed to the user. Backtrace printing is entirely unaffected by catch_unwind.

File: crates/wasmtime/src/config.rs (hunk 2 above)
Comment (TimonPost) [reply crates/wasmtime/src/config.rs:1587]: Unfortunately seems like self.compiler_config.ensure_setting_unset_or_given still needs a mutable borrow while tripple would be borrowed immutability.

File: crates/wasmtime/src/config.rs (hunk 3 above)
Comment (TimonPost) [reply crates/wasmtime/src/config.rs]: Updated!

File: crates/wasmtime/src/config.rs (hunk 2 above)
Comment (jameysharp) [reply crates/wasmtime/src/config.rs:1586]: Oops, you're right. We could fix that by moving the `ensure_setting_unset_or_given` method to `compiler_config.settings`, but cloning the triple isn't so bad either, so this is fine as-is.

**Review by jameysharp — APPROVED** [review]
Great, I think this is all set. Thank you!

**Comment by fitzgen** [comment]
> Does Wasmtime GC still require libunwind? I assumed the fast stack walking is used there now.

That is correct, we no longer require libunwind since the fast stack walking patches. Those docs are just outdated.

**Comment by jameysharp** [comment]
CI says that this PR breaks fuzzing because `crates/fuzzing/src/generators/config.rs` may randomly try to disable unwind info even on Windows. So we need the config generator to not do that on Windows.

I think these configs are never used for cross-compiling, so it should be good enough to use `cfg!(target_os = "windows")` rather than having to plumb the target through. But I'd appreciate if @fitzgen or @alexcrichton could check me on that.

Assuming that's true, then the `to_wasmtime` method in `crates/fuzzing/src/generators/config.rs` can just replace
```rust
.native_unwind_info(self.wasmtime.native_unwind_info)
```
with
```rust
.native_unwind_info(cfg!(target_os = "windows") || self.wasmtime.native_unwind_info)
```

**Comment by rockwotj** [comment]
@TimonPost would love to see this get merged, is there anything I can do to help?

**Comment by TimonPost** [comment]
Good one, i'll patch it up, should be ready in a bit!

**Comment by TimonPost** [comment]
Rebased and updated the PR. @fitzgen can you review this?

**Comment by rockwotj** [comment]
I'm not sure about the CI failures, but just figured I'd popin to say I'm using this patch with #6896 and it fixed some issues with C++ exceptions and wasmtime used in conjunction (there is a lock in libgcc that was having some contention issues).

**Review by jameysharp — APPROVED** [review]
This PR still looks good to me, thanks for fixing the Windows fuzzing issue!

The CI failure is probably because you accidentally changed `crates/wasi-nn/spec`. I suspect you need to run `git submodule update` to undo that, and then hopefully everything will be fine.
